### ISSUE TITLE
RNN ops input and output shapes are inconsistent with other ops

### ISSUE THREAD

**Issue by JamesAllingham** [onnx/onnx#2159]
Recurrent operators (`RNN`, `GRU`, `LSTM`) currently have the following input and output shapes:

```
X : T
    The input sequences packed (and potentially padded) into one 3-D tensor with the shape of `[seq_length, batch_size, input_size]`

initial_h (optional) : T
    Optional initial value of the hidden. If not specified - assumed to be 0. It has shape `[num_directions, batch_size, hidden_size]`.


Y (optional) : T
    A tensor that concats all the intermediate output values of the hidden. It has shape `[seq_length, num_directions, batch_size, hidden_size]`. 

Y_h (optional) : T
    The last output value of the hidden. It has shape `[num_directions, batch_size, hidden_size]`. 
```

Notice that the shapes of these inputs and outputs do not start with the batch size. On the other hand, *many* other ops, for example `Conv`, `ConvTranspose`, `AveragePool`, `MaxPool`, `SpaceToDepth`, `DepthToSpace`, `Multinomial`, `RoiAlign`, `BatchNorm`, etc., have inputs and outputs whose shapes start with the batch size. 

This is not ideal because it requires that extra `Transpose` operations be inserted when using the recurrent layers in conjunction with *batch first* layers. For example when exporting this simple net from Pytorch:

```
class Net(nn.Module):

    def __init__(self):
        super(Net, self).__init__()
        self.rnn1 = nn.RNN(10,10)
        self.fc1 = nn.Linear(10, 10)

    def forward(self, x):
        x, _ = self.rnn1(x)
        x = self.fc1(x)
        return x
```
we get an ONNX graph which contains an extra transpose:

```
%22 : Tensor, %23 : Float(1, 3, 10) = onnx::RNN[activations=["Tanh"], hidden_size=10](%input.1, %19, %20, %21, %17, %16), scope: Net/RNN[rnn1]
%24 : Float(10, 3, 10) = onnx::Squeeze[axes=[1]](%22), scope: Net/RNN[rnn1]
%25 : Float(10!, 10!) = onnx::Transpose[perm=[1, 0]](%fc1.weight), scope: Net/Linear[fc1]
%26 : Float(10, 3, 10) = onnx::MatMul(%24, %25), scope: Net/Linear[fc1]
```

The transpose would be unnecessary if the `RNN` output shape was batch first. 

Furthermore, most deep learning frameworks (including TF, Pytorch, and Keras) use have batch first implementations of `RNN`, `GRU`, and `LSTM`. Meaning that exporters from these frameworks have to insert Transpose ops, and the importers need to remove them, which creates extra work for their implementers.

Is there a strong reason not to make the recurrent ops batch first? 

The only reason that I could think of is that it is more efficient to do inference with a non-batch first implementation. I don't know if that is true. Even if it is, it seems like implementation details like that should not take precedence over simplicity and consistency. By consistency, I mean both with other ONNX ops and with the implementations of deep learning frameworks for which ONNX is aiming to be an interchange format. 

All of that said, maybe I am missing something here!

**Comment by ebarsoum** [comment]
Hi @JamesAllingham 

To efficiently writing recurrent network kernel you feed the kernel the batch of element for each sequence in the batch. So the preferred format for RNN and its family is [seq, batch, D], you iterate through the seq and passing [Batch, D] to the RNN (and its family) Kernel. TF implementation support [seq, batch, D] only, if you set time major to false, it will simply add transpose internally.

One option is to support both, in this case we can add time_major attribute similar to TF and other framework. Can you submit a PR for that?

Thanks,
Emad

**Comment by JamesAllingham** [comment]
Hi @ebarsoum - thanks for the response. I think that adding a time_major attribute is a good option. I will start working on a PR for that!

**Comment by jcwchen** [comment]
I think it should be resolved. Please reopen it if you still have questions. Thanks!
